contracts/src/WarpProtocolRegistry.sol

```diff
@@ -0,0 +1,195 @@
+    mapping(address => uint256) internal _addressToVersion;
+
+    /**
+     * @dev Emitted when a new protocol version is added to the registry.
+     */
+    event AddProtocolVersion(
+        uint256 indexed version,
+        address indexed protocolAddress
+    );
+
+    // Errors
+    error InvalidWarpMessage();
+    error InvalidSourceChainID();
+    error InvalidOriginSenderAddress();
+    error InvalidDestinationChainID();
+    error InvalidDestinationAddress();
+    error InvalidProtocolVersion();
+    error InvalidProtocolAddress();
+
+    /**
+     * @dev Initializes the contract by setting a `chainID` and `latestVersion`.
+     */
+    constructor(ProtocolRegistryEntry[] memory initialEntries) {
+        _latestVersion = 0;
+        _chainID = WARP_MESSENGER.getBlockchainID();
+
+        for (uint256 i = 0; i < initialEntries.length; i++) {
+            _addToRegistry(initialEntries[i]);
+        }
+    }
+
+    /**
+     * @dev Gets and verifies a warp out-of-band message, and adds the new protocol version
+     * address to the registry.
+     *
+     * Emits a {AddProtocolVersion} event when successful.
+     * Requirements:
+     *
+     * - a valid warp out-of-band message must be provided.
+     * - the version must be the increment of the latest version.
+     * - the protocol address must be a contract address.
+     */
+    function addProtocolVersion(uint32 messageIndex) external {
+        _addProtocolVersion(messageIndex);
+    }
+
+    /**
+     * @dev Gets the address of a protocol version.
+     * Requirements:
+     *
+     * - the version must be a valid version.
+     */
+    function getAddressFromVersion(
+        uint256 version
+    ) external view returns (address) {
+        return _getAddressFromVersion(version);
+    }
+
+    /**
+     * @dev Gets the version of the given `protocolAddress`.
+     * If `protocolAddress` is not a valid protocol address, returns 0, which is an invalid version.
+     */
+    function getVersionFromAddress(
+        address protocolAddress
+    ) external view returns (uint256) {
+        return _addressToVersion[protocolAddress];
+    }
+
+    /**
+     * @dev Gets the latest protocol version.
+     */
+    function getLatestVersion() external view returns (uint256) {
+        return _latestVersion;
+    }
+
+    /**
+     * @dev Gets and verifies for a warp out-of-band message, and adds the new protocol version
+     * addres to the registry.
+     */
+    function _addProtocolVersion(uint32 messageIndex) internal virtual {
+        // Get and validate for a warp out-of-band message.
+        (WarpMessage memory message, bool valid) = WARP_MESSENGER
+            .getVerifiedWarpMessage(messageIndex);
+        if (!valid) {
+            revert InvalidWarpMessage();
+        }
+        if (message.sourceChainID != _chainID) {
+            revert InvalidSourceChainID();
+        }
+        if (message.originSenderAddress != VALIDATORS_SOURCE_ADDRESS) {
```

Comment:
Can you add a comment here that this verifies the message is sent as an out-of-band Warp message?